File: packages/less/src/less/tree/media.test.js
import { describe, it, expect, vi, beforeEach } from 'vitest';
import Media from './media';

// Mock all dependencies to avoid circular imports and focus on Media behavior
vi.mock('./ruleset', () => ({
    default: vi.fn().mockImplementation(function (selectors, rules) {
        this.selectors = selectors;
        this.rules = rules;
        this.allowImports = false;
        this.functionRegistry = null;
        this.parent = null;
        this.debugInfo = null;
        this.eval = vi.fn().mockReturnValue(this);
        return this;
    })
}));

vi.mock('./value', () => ({
    default: vi.fn().mockImplementation(function (value) {
        this.value = Array.isArray(value) ? value : [value];
        this.eval = vi.fn().mockReturnValue(this);
        this.genCSS = vi.fn((_context, output) => {
            output.add('screen');
        });
        return this;
    })
}));

vi.mock('./selector', () => ({
    default: vi
        .fn()
        .mockImplementation(function (
            elements,
            extendList,
            condition,
            index,
            fileInfo
        ) {
            this.elements = elements || [];
            this.extendList = extendList;
            this.condition = condition;
            this._index = index;
            this._fileInfo = fileInfo;
            this.createEmptySelectors = vi.fn().mockReturnValue([
                {
                    elements: [],
                    extendList: null,
                    condition: null,
                    _index: index,
                    _fileInfo: fileInfo
                }
            ]);
            return this;
        })
}));

vi.mock('./atrule', () => ({
    default: function () {
        this.type = 'AtRule';
        this.outputRuleset = vi.fn((_context, output, rules) => {
            output.add(' { ');
            if (rules && rules.length > 0) {
                rules[0].genCSS && rules[0].genCSS(_context, output);
            }
            output.add(' }');
        });
        // Add Node methods that Media inherits
        this.copyVisibilityInfo = vi.fn((visibilityInfo) => {
            if (visibilityInfo) {
                this.visibilityBlocks = visibilityInfo.visibilityBlocks;
                this.nodeVisible = visibilityInfo.nodeVisible;
            }
        });
        this.setParent = vi.fn((nodes, _parent) => {
            if (Array.isArray(nodes)) {
                nodes.forEach((node) => {
                    if (node && typeof node === 'object') {
                        node.parent = _parent;
                    }
                });
            } else if (nodes && typeof nodes === 'object') {
                nodes.parent = _parent;
            }
        });
        this.visibilityInfo = vi.fn(() => ({
            visibilityBlocks: this.visibilityBlocks,
            nodeVisible: this.nodeVisible
        }));
        this.getIndex = vi.fn(() => this._index);
        this.fileInfo = vi.fn(() => this._fileInfo);
        return this;
    }
}));

vi.mock('./nested-at-rule', () => ({
    default: {
        isRulesetLike() {
            return true;
        },
        accept(visitor) {
            if (this.features) {
                this.features = visitor.visit(this.features);
            }
            if (this.rules) {
                this.rules = visitor.visitArray(this.rules);
            }
        },
        evalTop(/* _context */) {
            return this;
        },
        evalNested(/* _context */) {
            return this;
        },
        permute(arr) {
            if (arr.length === 0) return [];
            if (arr.length === 1) return arr[0];
            return [];
        },
        bubbleSelectors(/* _selectors */) {
            // Mock implementation
        }
    }
}));

describe('Media', () => {
    let mockContext;
    let mockOutput;
    let mockFileInfo;
    let mockVisibilityInfo;

    beforeEach(() => {
        vi.clearAllMocks();

        mockFileInfo = {
            filename: 'test.less',
            currentDirectory: '/test'
        };

        mockVisibilityInfo = {
            visibilityBlocks: 1,
            nodeVisible: true
        };

        mockContext = {
            mediaBlocks: [],
            mediaPath: [],
            frames: [
                {
                    functionRegistry: {
                        inherit: vi.fn().mockReturnValue({
                            get: vi.fn(),
                            add: vi.fn()
                        })
                    }
                }
            ]
        };

        mockOutput = {
            add: vi.fn()
        };
    });

    describe('constructor', () => {
        it('should create a Media instance with all parameters', () => {
            const value = [
                { type: 'Declaration', name: 'color', value: 'red' }
            ];
            const features = [{ type: 'Anonymous', value: 'screen' }];
            const index = 10;

            const media = new Media(
                value,
                features,
                index,
                mockFileInfo,
                mockVisibilityInfo
            );

            expect(media._index).toBe(index);
            expect(media._fileInfo).toBe(mockFileInfo);
            expect(media.features).toBeDefined();
            expect(media.rules).toHaveLength(1);
            expect(media.rules[0].allowImports).toBe(true);
            expect(media.allowRoot).toBe(true);
        });

        it('should handle null/undefined value', () => {
            const features = [{ type: 'Anonymous', value: 'screen' }];

            const media1 = new Media(
                null,
                features,
                0,
                mockFileInfo,
                mockVisibilityInfo
            );
            const media2 = new Media(
                undefined,
                features,
                0,
                mockFileInfo,
                mockVisibilityInfo
            );

            expect(media1.rules).toHaveLength(1);
            expect(media2.rules).toHaveLength(1);
        });

        it('should handle empty features', () => {
            const value = [
                { type: 'Declaration', name: 'color', value: 'red' }
            ];

            const media = new Media(
                value,
                [],
                0,
                mockFileInfo,
                mockVisibilityInfo
            );

            expect(media.features).toBeDefined();
            expect(media.rules).toHaveLength(1);
        });

        it('should handle missing index and fileInfo', () => {
            const value = [
                { type: 'Declaration', name: 'color', value: 'red' }
            ];
            const features = [{ type: 'Anonymous', value: 'screen' }];

            const media = new Media(value, features);

            expect(media._index).toBeUndefined();
            expect(media._fileInfo).toBeUndefined();
            expect(media.rules).toHaveLength(1);
        });

        it('should handle missing visibilityInfo', () => {
            const value = [
                { type: 'Declaration', name: 'color', value: 'red' }
            ];
            const features = [{ type: 'Anonymous', value: 'screen' }];

            const media = new Media(value, features, 0, mockFileInfo);

            expect(media.rules).toHaveLength(1);
            expect(media.allowRoot).toBe(true);
        });

        it('should set parent relationships correctly', () => {
            const value = [
                { type: 'Declaration', name: 'color', value: 'red' }
            ];
            const features = [{ type: 'Anonymous', value: 'screen' }];

            const media = new Media(
                value,
                features,
                0,
                mockFileInfo,
                mockVisibilityInfo
            );

            // Check that setParent was called via the constructor
            expect(media.features).toBeDefined();
            expect(media.rules).toBeDefined();
        });
    });

    describe('type property', () => {
        it('should have type "Media"', () => {
            const media = new Media([], [], 0, mockFileInfo);
            expect(media.type).toBe('Media');
        });
    });

    describe('genCSS method', () => {
        it('should generate CSS with @media rule', () => {
            const value = [
                { type: 'Declaration', name: 'color', value: 'red' }
            ];
            const features = [{ type: 'Anonymous', value: 'screen' }];
            const media = new Media(value, features, 5, mockFileInfo);

            media.genCSS(mockContext, mockOutput);

            expect(mockOutput.add).toHaveBeenCalledWith(
                '@media ',
                mockFileInfo,
                5
            );
            expect(media.features.genCSS).toHaveBeenCalledWith(
                mockContext,
                mockOutput
            );
            expect(media.outputRuleset).toHaveBeenCalledWith(
                mockContext,
                mockOutput,
                media.rules
            );
        });

        it('should handle context without fileInfo', () => {
            const media = new Media([], [], 0);

            media.genCSS({}, mockOutput);

            expect(mockOutput.add).toHaveBeenCalledWith(
                '@media ',
                undefined,
                0
            );
        });

        it('should handle null context', () => {
            const media = new Media([], [], 0, mockFileInfo);

            expect(() => media.genCSS(null, mockOutput)).not.toThrow();
            expect(mockOutput.add).toHaveBeenCalledWith(
                '@media ',
                mockFileInfo,
                0
            );
        });

        it('should handle different output objects', () => {
            const customOutput = { add: vi.fn() };
            const media = new Media([], [], 0, mockFileInfo);

            media.genCSS(mockContext, customOutput);

            expect(customOutput.add).toHaveBeenCalledWith(
                '@media ',
                mockFileInfo,
                0
            );
        });
    });

    describe('eval method', () => {
        beforeEach(() => {
            mockContext = {
                mediaBlocks: undefined,
                mediaPath: undefined,
                frames: [
                    {
                        functionRegistry: {
                            inherit: vi.fn().mockReturnValue({
                                get: vi.fn(),
                                add: vi.fn()
                            })
                        }
                    }
                ]
            };
        });

        it('should initialize mediaBlocks and mediaPath when not present', () => {
            const media = new Media([], [], 0, mockFileInfo);

            const result = media.eval(mockContext);

            expect(mockContext.mediaBlocks).toBeDefined();
            expect(mockContext.mediaBlocks).toHaveLength(1); // Media adds itself to mediaBlocks
            expect(mockContext.mediaPath).toBeDefined();
            expect(result).toBeDefined();
        });

        it('should handle existing mediaBlocks and mediaPath', () => {
            mockContext.mediaBlocks = [{ type: 'Media' }];
            mockContext.mediaPath = [{ type: 'Media' }];

            const media = new Media([], [], 0, mockFileInfo);
            const result = media.eval(mockContext);

            expect(mockContext.mediaBlocks).toHaveLength(2); // original + new media
            expect(result).toBeDefined();
        });

        it('should preserve debugInfo when present', () => {
            const media = new Media([], [], 0, mockFileInfo);
            media.debugInfo = { lineNumber: 5, fileName: 'test.less' };

            const result = media.eval(mockContext);

            expect(result.debugInfo).toEqual({
                lineNumber: 5,
                fileName: 'test.less'
            });
        });

        it('should evaluate features', () => {
            const media = new Media(
                [],
                [{ type: 'Anonymous', value: 'screen' }],
                0,
                mockFileInfo
            );

            media.eval(mockContext);

            expect(media.features.eval).toHaveBeenCalledWith(mockContext);
        });

        it('should handle context with frames', () => {
            const media = new Media(
                [{ type: 'Declaration' }],
                [],
                0,
                mockFileInfo
            );

            const result = media.eval(mockContext);

            expect(mockContext.frames).toHaveLength(1); // should be restored after eval
            expect(result).toBeDefined();
        });

        it('should return evalTop result when mediaPath is empty', () => {
            const media = new Media([], [], 0, mockFileInfo);
            mockContext.mediaBlocks = [];
            mockContext.mediaPath = [];

            const result = media.eval(mockContext);

            // mediaPath should be popped back to empty, triggering evalTop
            expect(result).toBeDefined();
        });

        it('should return evalNested result when mediaPath is not empty', () => {
            const media = new Media([], [], 0, mockFileInfo);
            mockContext.mediaBlocks = [];
            mockContext.mediaPath = [{ type: 'Media' }]; // Pre-existing media in path

            const result = media.eval(mockContext);

            expect(result).toBeDefined();
        });

        it('should handle rules evaluation', () => {
            // Create a media instance and access its rules to test evaluation
            const media = new Media(
                [{ type: 'Declaration' }],
                [],
                0,
                mockFileInfo
            );

            // The eval method should work and return a result
            const result = media.eval(mockContext);

            expect(result).toBeDefined();
            expect(media.rules).toHaveLength(1);
            expect(media.rules[0].eval).toHaveBeenCalledWith(mockContext);
        });

        it('should manage context frames correctly', () => {
            const originalFramesLength = mockContext.frames.length;
            const media = new Media(
                [{ type: 'Declaration' }],
                [],
                0,
                mockFileInfo
            );

            media.eval(mockContext);

            // Frames should be restored to original length
            expect(mockContext.frames).toHaveLength(originalFramesLength);
        });

        it('should handle function registry inheritance', () => {
            const media = new Media([], [], 0, mockFileInfo);

            media.eval(mockContext);

            expect(
                mockContext.frames[0].functionRegistry.inherit
            ).toHaveBeenCalled();
        });
    });

    describe('inheritance and prototype chain', () => {
        it('should inherit from AtRule', () => {
            const media = new Media([], [], 0, mockFileInfo);
            expect(media.type).toBe('Media');
            expect(typeof media.outputRuleset).toBe('function');
        });

        it('should include NestableAtRulePrototype methods', () => {
            const media = new Media([], [], 0, mockFileInfo);
            expect(typeof media.isRulesetLike).toBe('function');
            expect(media.isRulesetLike()).toBe(true);
        });
    });

    describe('edge cases and error handling', () => {
        it('should handle complex features array', () => {
            const complexFeatures = [
                { type: 'Anonymous', value: 'screen' },
                { type: 'Anonymous', value: 'and' },
                { type: 'Expression', value: '(max-width: 768px)' }
            ];

            const media = new Media([], complexFeatures, 0, mockFileInfo);

            expect(media.features).toBeDefined();
            expect(media.rules).toHaveLength(1);
        });

        it('should handle complex value array', () => {
            const complexValue = [
                { type: 'Declaration', name: 'color', value: 'red' },
                { type: 'Declaration', name: 'background', value: 'blue' },
                { type: 'Ruleset', selectors: [], rules: [] }
            ];

            const media = new Media(complexValue, [], 0, mockFileInfo);

            expect(media.rules).toHaveLength(1);
            expect(media.rules[0].allowImports).toBe(true);
        });

        it('should handle zero index', () => {
            const media = new Media([], [], 0, mockFileInfo);
            expect(media._index).toBe(0);
        });

        it('should handle negative index', () => {
            const media = new Media([], [], -1, mockFileInfo);
            expect(media._index).toBe(-1);
        });

        it('should handle large index', () => {
            const media = new Media([], [], 999999, mockFileInfo);
            expect(media._index).toBe(999999);
        });

        it('should handle empty fileInfo object', () => {
            const emptyFileInfo = {};
            const media = new Media([], [], 0, emptyFileInfo);
            expect(media._fileInfo).toBe(emptyFileInfo);
        });

        it('should handle fileInfo with minimal properties', () => {
            const minimalFileInfo = { filename: 'test.less' };
            const media = new Media([], [], 0, minimalFileInfo);
            expect(media._fileInfo).toBe(minimalFileInfo);
        });
    });

    describe('context handling in eval', () => {
        it('should handle context without frames', () => {
            const contextWithoutFrames = {
                mediaBlocks: undefined,
                mediaPath: undefined
            };

            const media = new Media([], [], 0, mockFileInfo);

            expect(() => media.eval(contextWithoutFrames)).toThrow();
        });

        it('should handle context with empty frames array', () => {
            const contextWithEmptyFrames = {
                mediaBlocks: undefined,
                mediaPath: undefined,
                frames: []
            };

            const media = new Media([], [], 0, mockFileInfo);

            expect(() => media.eval(contextWithEmptyFrames)).toThrow();
        });

        it('should handle context with multiple frames', () => {
            const contextWithMultipleFrames = {
                mediaBlocks: undefined,
                mediaPath: undefined,
                frames: [
                    {
                        functionRegistry: {
                            inherit: vi.fn().mockReturnValue({})
                        }
                    },
                    {
                        functionRegistry: {
                            inherit: vi.fn().mockReturnValue({})
                        }
                    },
                    {
                        functionRegistry: {
                            inherit: vi.fn().mockReturnValue({})
                        }
                    }
                ]
            };

            const media = new Media([], [], 0, mockFileInfo);
            const result = media.eval(contextWithMultipleFrames);

            expect(result).toBeDefined();
            expect(contextWithMultipleFrames.frames).toHaveLength(3); // Should be restored
        });
    });

    describe('visibility and parent handling', () => {
        it('should copy visibility info correctly', () => {
            const visibilityInfo = {
                visibilityBlocks: 2,
                nodeVisible: false
            };

            const media = new Media([], [], 0, mockFileInfo, visibilityInfo);

            // The visibility info should be copied to the media instance
            expect(media.visibilityBlocks).toBe(2);
            expect(media.nodeVisible).toBe(false);
        });

        it('should handle null visibility info', () => {
            const media = new Media([], [], 0, mockFileInfo, null);

            expect(media.rules).toHaveLength(1);
        });

        it('should handle undefined visibility info', () => {
            const media = new Media([], [], 0, mockFileInfo, undefined);

            expect(media.rules).toHaveLength(1);
        });
    });
});

File: packages/less/src/less/less_go/media_test.go
package less_go

import (
	"fmt"
	"testing"
)

func TestMediaConstructor(t *testing.T) {
	t.Run("should create a Media instance with all parameters", func(t *testing.T) {
		value := []any{
			map[string]any{"type": "Declaration", "name": "color", "value": "red"},
		}
		features := []any{map[string]any{"type": "Anonymous", "value": "screen"}}
		index := 10
		mockFileInfo := map[string]any{
			"filename":         "test.less",  
			"currentDirectory": "/test",
		}
		mockVisibilityInfo := map[string]any{
			"visibilityBlocks": 1,
			"nodeVisible":      true,
		}

		media := NewMedia(value, features, index, mockFileInfo, mockVisibilityInfo)

		if media.GetIndex() != index {
			t.Errorf("Expected index %d, got %d", index, media.GetIndex())
		}
		if media.FileInfo() == nil {
			t.Error("Expected fileInfo to be set")
		}
		if media.Features == nil {
			t.Error("Expected features to be defined")
		}
		if len(media.Rules) != 1 {
			t.Errorf("Expected 1 rule, got %d", len(media.Rules))
		}
		if ruleset, ok := media.Rules[0].(*Ruleset); ok {
			if !ruleset.AllowImports {
				t.Error("Expected allowImports to be true")
			}
		} else {
			t.Error("Expected first rule to be a Ruleset")
		}
		if !media.AllowRoot {
			t.Error("Expected allowRoot to be true")
		}
	})

	t.Run("should handle null/undefined value", func(t *testing.T) {
		features := []any{map[string]any{"type": "Anonymous", "value": "screen"}}
		mockFileInfo := map[string]any{
			"filename":         "test.less",
			"currentDirectory": "/test",
		}
		mockVisibilityInfo := map[string]any{
			"visibilityBlocks": 1,
			"nodeVisible":      true,
		}

		media1 := NewMedia(nil, features, 0, mockFileInfo, mockVisibilityInfo)
		media2 := NewMedia(nil, features, 0, mockFileInfo, mockVisibilityInfo)

		if len(media1.Rules) != 1 {
			t.Errorf("Expected 1 rule for media1, got %d", len(media1.Rules))
		}
		if len(media2.Rules) != 1 {
			t.Errorf("Expected 1 rule for media2, got %d", len(media2.Rules))
		}
	})

	t.Run("should handle empty features", func(t *testing.T) {
		value := []any{
			map[string]any{"type": "Declaration", "name": "color", "value": "red"},
		}
		mockFileInfo := map[string]any{
			"filename":         "test.less",
			"currentDirectory": "/test",
		}
		mockVisibilityInfo := map[string]any{
			"visibilityBlocks": 1,
			"nodeVisible":      true,
		}

		media := NewMedia(value, []any{}, 0, mockFileInfo, mockVisibilityInfo)

		if media.Features == nil {
			t.Error("Expected features to be defined")
		}
		if len(media.Rules) != 1 {
			t.Errorf("Expected 1 rule, got %d", len(media.Rules))
		}
	})

	t.Run("should handle missing index and fileInfo", func(t *testing.T) {
		value := []any{
			map[string]any{"type": "Declaration", "name": "color", "value": "red"},
		}
		features := []any{map[string]any{"type": "Anonymous", "value": "screen"}}

		media := NewMedia(value, features, 0, nil, nil)

		if len(media.Rules) != 1 {
			t.Errorf("Expected 1 rule, got %d", len(media.Rules))
		}
	})

	t.Run("should handle missing visibilityInfo", func(t *testing.T) {
		value := []any{
			map[string]any{"type": "Declaration", "name": "color", "value": "red"},
		}
		features := []any{map[string]any{"type": "Anonymous", "value": "screen"}}
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}

		media := NewMedia(value, features, 0, mockFileInfo, nil)

		if len(media.Rules) != 1 {
			t.Errorf("Expected 1 rule, got %d", len(media.Rules))
		}
		if !media.AllowRoot {
			t.Error("Expected allowRoot to be true")
		}
	})
}

func TestMediaType(t *testing.T) {
	t.Run("should have type Media", func(t *testing.T) {
		mockFileInfo := map[string]any{"filename": "test.less"}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)
		
		if media.GetType() != "Media" {
			t.Errorf("Expected type 'Media', got '%s'", media.GetType())
		}
	})
}

func TestMediaGenCSS(t *testing.T) {
	// Create output for testing
	var outputContent []string
	output := &CSSOutput{
		Add: func(chunk any, fileInfo any, index any) {
			if chunk != nil {
				// Handle different types of chunks
				switch v := chunk.(type) {
				case string:
					outputContent = append(outputContent, v)
				default:
					// Convert other types to string representation
					outputContent = append(outputContent, fmt.Sprintf("%v", v))
				}
			}
		},
		IsEmpty: func() bool {
			return len(outputContent) == 0
		},
	}
	
	t.Run("should generate CSS with @media rule", func(t *testing.T) {
		value := []any{
			map[string]any{"type": "Declaration", "name": "color", "value": "red"},
		}
		features := []any{map[string]any{"type": "Anonymous", "value": "screen"}}
		mockFileInfo := map[string]any{
			"filename":         "test.less",
			"currentDirectory": "/test",
		}
		media := NewMedia(value, features, 5, mockFileInfo, nil)
		
		mockContext := map[string]any{}

		// The GenCSS method should not panic
		media.GenCSS(mockContext, output)
		
		// Check that some content was added to the output
		if output.IsEmpty() {
			t.Error("Expected some output to be generated")
		}
	})

	t.Run("should handle context without fileInfo", func(t *testing.T) {
		media := NewMedia([]any{}, []any{}, 0, nil, nil)

		mockContext := map[string]any{}
		
		// Should not panic
		media.GenCSS(mockContext, output)
	})

	t.Run("should handle nil context", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)

		// Should not panic
		media.GenCSS(nil, output)
	})
}

func TestMediaEval(t *testing.T) {
	t.Run("should initialize mediaBlocks and mediaPath when not present", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename":         "test.less",
			"currentDirectory": "/test",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)

		mockContext := map[string]any{
			"frames": []any{
				&Ruleset{
					FunctionRegistry: map[string]any{},
				},
			},
		}

		result, err := media.Eval(mockContext)
		if err != nil {
			t.Errorf("Unexpected error: %v", err)
		}
		if result == nil {
			t.Error("Expected result to be defined")
		}
		
		// Check that mediaBlocks and mediaPath were initialized
		if _, ok := mockContext["mediaBlocks"]; !ok {
			t.Error("Expected mediaBlocks to be defined in context")
		}
		if _, ok := mockContext["mediaPath"]; !ok {
			t.Error("Expected mediaPath to be defined in context")
		}
	})

	t.Run("should handle existing mediaBlocks and mediaPath", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)
		
		mockContext := map[string]any{
			"mediaBlocks": []any{map[string]any{"type": "Media"}},
			"mediaPath":   []any{map[string]any{"type": "Media"}},
			"frames": []any{
				&Ruleset{
					FunctionRegistry: map[string]any{},
				},
			},
		}

		result, err := media.Eval(mockContext)
		if err != nil {
			t.Errorf("Unexpected error: %v", err)
		}
		if result == nil {
			t.Error("Expected result to be defined")
		}
	})

	t.Run("should preserve debugInfo when present", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)
		media.DebugInfo = map[string]any{"lineNumber": 5, "fileName": "test.less"}

		mockContext := map[string]any{
			"frames": []any{
				&Ruleset{
					FunctionRegistry: map[string]any{},
				},
			},
		}

		result, err := media.Eval(mockContext)
		if err != nil {
			t.Errorf("Unexpected error: %v", err)
		}
		
		if resultMedia, ok := result.(*Media); ok {
			if resultMedia.DebugInfo == nil {
				t.Error("Expected debugInfo to be preserved")
			}
		}
	})

	t.Run("should handle context with frames", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{map[string]any{"type": "Declaration"}}, []any{}, 0, mockFileInfo, nil)

		mockContext := map[string]any{
			"frames": []any{
				&Ruleset{
					FunctionRegistry: map[string]any{},
				},
			},
		}

		result, err := media.Eval(mockContext)
		if err != nil {
			t.Errorf("Unexpected error: %v", err)
		}
		if result == nil {
			t.Error("Expected result to be defined")
		}
		
		// Should restore frames to original length
		if frames, ok := mockContext["frames"].([]any); ok {
			if len(frames) != 1 {
				t.Errorf("Expected frames to be restored to length 1, got %d", len(frames))
			}
		}
	})

	t.Run("should return evalTop result when mediaPath is empty", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)
		
		mockContext := map[string]any{
			"mediaBlocks": []any{},
			"mediaPath":   []any{},
			"frames": []any{
				&Ruleset{
					FunctionRegistry: map[string]any{},
				},
			},
		}

		result, err := media.Eval(mockContext)
		if err != nil {
			t.Errorf("Unexpected error: %v", err)
		}
		if result == nil {
			t.Error("Expected result to be defined")
		}
	})

	t.Run("should return evalNested result when mediaPath is not empty", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)
		
		mockContext := map[string]any{
			"mediaBlocks": []any{},
			"mediaPath":   []any{map[string]any{"type": "Media"}}, // Pre-existing media in path
			"frames": []any{
				&Ruleset{
					FunctionRegistry: map[string]any{},
				},
			},
		}

		result, err := media.Eval(mockContext)
		if err != nil {
			t.Errorf("Unexpected error: %v", err)
		}
		if result == nil {
			t.Error("Expected result to be defined")
		}
	})
}

func TestMediaInheritanceAndPrototypeChain(t *testing.T) {
	t.Run("should inherit from AtRule", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)
		
		if media.GetType() != "Media" {
			t.Errorf("Expected type 'Media', got %s", media.GetType())
		}
		// Check that it has OutputRuleset method from AtRule (method exists via embedding)
		// Testing that we can access the inherited method without errors
		if media.AtRule == nil {
			t.Error("Expected AtRule to be embedded")
		}
	})

	t.Run("should include NestableAtRulePrototype methods", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)
		
		if !media.IsRulesetLike() {
			t.Error("Expected IsRulesetLike to return true")
		}
	})
}

func TestMediaEdgeCasesAndErrorHandling(t *testing.T) {
	t.Run("should handle complex features array", func(t *testing.T) {
		complexFeatures := []any{
			map[string]any{"type": "Anonymous", "value": "screen"},
			map[string]any{"type": "Anonymous", "value": "and"},
			map[string]any{"type": "Expression", "value": "(max-width: 768px)"},
		}
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}

		media := NewMedia([]any{}, complexFeatures, 0, mockFileInfo, nil)

		if media.Features == nil {
			t.Error("Expected features to be defined")
		}
		if len(media.Rules) != 1 {
			t.Errorf("Expected 1 rule, got %d", len(media.Rules))
		}
	})

	t.Run("should handle complex value array", func(t *testing.T) {
		complexValue := []any{
			map[string]any{"type": "Declaration", "name": "color", "value": "red"},
			map[string]any{"type": "Declaration", "name": "background", "value": "blue"},
			map[string]any{"type": "Ruleset", "selectors": []any{}, "rules": []any{}},
		}
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}

		media := NewMedia(complexValue, []any{}, 0, mockFileInfo, nil)

		if len(media.Rules) != 1 {
			t.Errorf("Expected 1 rule, got %d", len(media.Rules))
		}
		if ruleset, ok := media.Rules[0].(*Ruleset); ok {
			if !ruleset.AllowImports {
				t.Error("Expected allowImports to be true")
			}
		}
	})

	t.Run("should handle zero index", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)
		
		if media.GetIndex() != 0 {
			t.Errorf("Expected index 0, got %d", media.GetIndex())
		}
	})

	t.Run("should handle negative index", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, -1, mockFileInfo, nil)
		
		if media.GetIndex() != -1 {
			t.Errorf("Expected index -1, got %d", media.GetIndex())
		}
	})

	t.Run("should handle large index", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 999999, mockFileInfo, nil)
		
		if media.GetIndex() != 999999 {
			t.Errorf("Expected index 999999, got %d", media.GetIndex())
		}
	})

	t.Run("should handle empty fileInfo object", func(t *testing.T) {
		emptyFileInfo := map[string]any{}
		media := NewMedia([]any{}, []any{}, 0, emptyFileInfo, nil)
		
		if media.FileInfo() == nil {
			t.Error("Expected fileInfo to be set")
		}
	})

	t.Run("should handle fileInfo with minimal properties", func(t *testing.T) {
		minimalFileInfo := map[string]any{"filename": "test.less"}
		media := NewMedia([]any{}, []any{}, 0, minimalFileInfo, nil)
		
		if media.FileInfo() == nil {
			t.Error("Expected fileInfo to be set")
		}
		if filename, ok := media.FileInfo()["filename"]; !ok || filename != "test.less" {
			t.Error("Expected filename to be 'test.less'")
		}
	})
}

func TestMediaContextHandlingInEval(t *testing.T) {
	t.Run("should handle context without frames", func(t *testing.T) {
		contextWithoutFrames := map[string]any{
			"mediaBlocks": nil,
			"mediaPath":   nil,
		}
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}

		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)

		// Should return error due to missing frames
		_, err := media.Eval(contextWithoutFrames)
		if err == nil {
			t.Error("Expected error due to missing frames")
		}
	})

	t.Run("should handle context with empty frames array", func(t *testing.T) {
		contextWithEmptyFrames := map[string]any{
			"mediaBlocks": nil,
			"mediaPath":   nil,
			"frames":      []any{},
		}
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}

		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)

		// Should handle empty frames gracefully
		result, err := media.Eval(contextWithEmptyFrames)
		if err != nil {
			t.Errorf("Unexpected error: %v", err)
		}
		if result == nil {
			t.Error("Expected result to be defined")
		}
	})

	t.Run("should handle context with multiple frames", func(t *testing.T) {
		contextWithMultipleFrames := map[string]any{
			"mediaBlocks": nil,
			"mediaPath":   nil,
			"frames": []any{
				&Ruleset{FunctionRegistry: map[string]any{}},
				&Ruleset{FunctionRegistry: map[string]any{}},
				&Ruleset{FunctionRegistry: map[string]any{}},
			},
		}
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}

		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)
		result, err := media.Eval(contextWithMultipleFrames)
		
		if err != nil {
			t.Errorf("Unexpected error: %v", err)
		}
		if result == nil {
			t.Error("Expected result to be defined")
		}
		
		// Should be restored to original length
		if frames, ok := contextWithMultipleFrames["frames"].([]any); ok {
			if len(frames) != 3 {
				t.Errorf("Expected frames to be restored to length 3, got %d", len(frames))
			}
		}
	})
}

func TestMediaVisibilityAndParentHandling(t *testing.T) {
	t.Run("should copy visibility info correctly", func(t *testing.T) {
		visibilityInfo := map[string]any{
			"visibilityBlocks": 2,
			"nodeVisible":      false,
		}
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}

		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, visibilityInfo)

		// The visibility info should be available through VisibilityInfo()
		if media.VisibilityInfo() == nil {
			t.Error("Expected visibility info to be copied")
		}
	})

	t.Run("should handle nil visibility info", func(t *testing.T) {
		mockFileInfo := map[string]any{
			"filename": "test.less",
		}
		media := NewMedia([]any{}, []any{}, 0, mockFileInfo, nil)

		if len(media.Rules) != 1 {
			t.Errorf("Expected 1 rule, got %d", len(media.Rules))
		}
	})
} 